### PR TITLE
Fix the HttpSys client disconnect race condition 

### LINKED ISSUE #12194: Headers already sent in HttpSys response caching
We are seeing occasional "Headers already sent" errors triggered from the HttpSys code that enables response caching:
```
ProcessRequestAsync System.InvalidOperationException: Headers already sent.
   at Microsoft.AspNetCore.Server.HttpSys.Response.CheckResponseStarted()
   at Microsoft.AspNetCore.Server.HttpSys.FeatureContext.ConsiderEnablingResponseCache()
   at Microsoft.AspNetCore.Server.HttpSys.FeatureContext.<OnResponseStart>d__131.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.HttpSys.MessagePump.<ProcessRequestAsync>d__23.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.Server.HttpSys.MessagePump.<ProcessRequestAsync>d__23.MoveNext()
```

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #12194 If the client disconnects before the request body is finished that causes the whole request and response to be aborted. This PR fixes a state tracking issue where that Abort would set the response state to Closed and cause misleading errors for the application. The application may be trying to handle the error in a generic way such as setting a 500 status code. However, because the response is now Closed, that triggers a HasStarted error.
https://github.com/dotnet/aspnetcore/blob/e1aeac809cb6dae3e95b9b896a1fc0e7d48b1a63/src/Servers/HttpSys/src/RequestProcessing/Response.cs#L60-L66
https://github.com/dotnet/aspnetcore/blob/e1aeac809cb6dae3e95b9b896a1fc0e7d48b1a63/src/Servers/HttpSys/src/RequestProcessing/Response.cs#L197-L201
https://github.com/dotnet/aspnetcore/blob/e1aeac809cb6dae3e95b9b896a1fc0e7d48b1a63/src/Servers/HttpSys/src/RequestProcessing/Response.cs#L230-L233

The fix is to track the client abort state independent of the application response state. E.g. the client can abort the request & response, but it doesn't change the response state to closed, the application can keep modifying the response independently. This avoids race conditions on disconnect and gives the app a consistent experience. The app can still explicitly check for disconnects from RequestAborted.

### CODE REVIEW

File: src/Servers/HttpSys/src/RequestProcessing/RequestStreamAsyncResult.cs [hunk 1]
```diff
@@ -135,6 +135,8 @@ internal void Complete(int read, uint errorCode = UnsafeNclNativeMethods.ErrorCo
 
         internal void Fail(Exception ex)
         {
+            Dispose();
+            _requestStream.Abort();
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/RequestProcessing/RequestStreamAsyncResult.cs:139]: This was moved to further limit races between the Abort logic and the ReadAsync Task completion.

Comment (JamesNK) [reply src/Servers/HttpSys/src/RequestProcessing/RequestStreamAsyncResult.cs:139]: Add a comment to the code saying this?

File: src/Servers/HttpSys/test/FunctionalTests/ResponseTests.cs [hunk 2]
```diff
@@ -210,6 +213,71 @@ public async Task Response_OnStartingThrowsAfterWrite_WriteThrowsAndStillCallsOn
...
+                    await readTask;
+                    readCompleted.SetException(new InvalidOperationException("The read wasn't supposed to succeed"));
+                    return;
+                }
+                catch (IOException)
+                {
+                }
+
+                // Let the Abort logic finish.
+                await Task.Yield();
```
Comment (JamesNK) [inline src/Servers/HttpSys/test/FunctionalTests/ResponseTests.cs:236]: This doesn't seem reliable.

What about:
1. Adding a TCS
2. Using `httpContext.RequestAbort` callback to set the TCS as complete
3. Awaiting TCS here.

That ensures the following logic always happens after the server has received the client disconnection.

Comment (Tratcher) [reply src/Servers/HttpSys/test/FunctionalTests/ResponseTests.cs:236]: With the re-ordered code in Fail I might be able to remove this entirely.

Comment (Tratcher) [reply src/Servers/HttpSys/test/FunctionalTests/ResponseTests.cs:236]: To clarify, this Yield was only here to make the original race condition more prominent so I could verify it was fixed, the test did not require it to pass.

**Review by avparuch — APPROVED** [review]

**Review by wtgodbe — APPROVED** [review]
Seems reasonable to me

**Pull request merged**